### PR TITLE
Node agent workload implementation

### PR DESCRIPTION
**Pull request by alexis-langlet**
Define workload trait and implement it for docker workloads.

Allow to create and stop docker workloads

### CODE REVIEW

File: node-agent/workload_manager/src/workload_manager/workload/container.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+use proto::agent::Instance;
+
+pub struct Container {
+    id: String,
+}
+
+impl Container {
+    /**
+     * Removes a container
+     */
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/container.rs:20]: Please use the following comment format:
```rust
//
// Removes a container
// 
```

File: node-agent/workload_manager/src/workload_manager/workload/container.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+impl Workload for Container {
+    fn get_id(&self) -> String {
+        self.id.to_string()
+    }
+
+    /**
+     * Create a new workload and start it
+     */
+    async fn new(instance: Instance) -> Self {
+        let docker = Docker::connect_with_socket_defaults().unwrap();
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/container.rs:46]: You're including a bunch of `unwrap()`. Please try to define and propagate errors properly instead.

File: node-agent/workload_manager/src/workload_manager/workload/mod.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+use workload_trait::Workload;
+
+mod container;
+pub mod errors;
+pub mod workload_trait;
+
+pub async fn create(instance: Instance) -> Result<impl Workload, WorkloadError> {
+    match instance.r#type {
+        // 0 for Container
+        0 => Ok(container::Container::new(instance).await),
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/mod.rs:12]: I'm hoping there's an enum for the instance type?

File: node-agent/workload_manager/src/workload_manager/workload/workload_trait.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+use proto::agent::Instance;
+#[tonic::async_trait]
+pub trait Workload {
+    /**
+     * Create a new workload and start it
+     */
+    async fn new(instance: Instance) -> Self;
+
+    fn get_id(&self) -> String;
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/workload_trait.rs:9]: `id`, not `get_id`.

File: node-agent/workload_manager/src/workload_manager/workload/workload_trait.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    /**
+     * Gracefully stops a workload
+     */
+    async fn stop(&self);
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/workload_trait.rs:14]: `stop` and `kill` should return a `Result`.

File: node-agent/workload_manager/src/workload_manager/workload/errors.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+use std::error::Error;
+use std::fmt;
+
+#[derive(Debug)]
+pub struct WorkloadError {
+    details: String,
+}
+
```
Comment (sameo) [inline node-agent/workload_manager/src/workload_manager/workload/errors.rs:8]: You should use the `anyhow` crate instead.

Comment (alexis-langlet) [reply node-agent/workload_manager/src/workload_manager/workload/errors.rs:8]: Sorry, I don't understand. 
Does this mean that I'm supposed to change the #[derive(Debug)] for something from anyhow crate (but I didn't see anything that looks like it in the doc) or should I just remove WorkloadError struct and just use anyhow basics errors instead ?

Comment (sameo) [reply node-agent/workload_manager/src/workload_manager/workload/errors.rs:8]: `WorkloadError` is only a string, the `anyhow` create already define an error type that does exactly this, implements `Error`, etc. So you could remove `WorkloadError` and simply use `anyhow` instead.

File: node-agent/workload_manager/src/workload_manager/workload/mod.rs (hunk 3 above)
Comment (alexis-langlet) [reply node-agent/workload_manager/src/workload_manager/workload/mod.rs]: Yes, sorry...

File: node-agent/workload_manager/src/workload_manager/workload/errors.rs (hunk 6 above)
Comment (alexis-langlet) [reply node-agent/workload_manager/src/workload_manager/workload/errors.rs]: Ok, thank you, that's a lot more practical indeed !

**Review by sameo — APPROVED** [review]

**Pull request merged**